File: library/src/test/java/me/drakeet/multitype/MultiTypeAdapterTest.java
package me.drakeet.multitype;

import android.content.Context;
import android.support.annotation.NonNull;
import android.view.LayoutInflater;
import android.view.ViewGroup;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.robolectric.RobolectricTestRunner;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertTrue;
import static org.mockito.ArgumentMatchers.anyList;
import static org.mockito.ArgumentMatchers.anyString;
import static org.mockito.ArgumentMatchers.eq;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;
import static org.mockito.MockitoAnnotations.initMocks;

/**
 * @author drakeet
 */
@RunWith(RobolectricTestRunner.class)
public class MultiTypeAdapterTest {

  @Mock
  private ViewGroup parent;
  @Mock
  private Context context;
  @Mock
  private TestItemViewBinder mockedItemViewBinder;
  @Mock
  private LayoutInflater inflater;

  private TestItemViewBinder itemViewBinder = new TestItemViewBinder();


  @Before
  public void setUp() throws Exception {
    initMocks(this);
    when(parent.getContext()).thenReturn(context);
    when(context.getSystemService(anyString())).thenReturn(inflater);
  }


  @Test
  public void shouldReturnOriginalItems() {
    List list = new ArrayList();
    MultiTypeAdapter adapter = new MultiTypeAdapter(list);
    assertEquals(list, adapter.getItems());
  }


  @Test
  public void shouldReturnEmptyItemsWithDefaultConstructor() {
    MultiTypeAdapter adapter = new MultiTypeAdapter();
    assertTrue(adapter.getItems().isEmpty());
  }


  @Test
  public void shouldOverrideRegisteredBinder() {
    MultiTypeAdapter adapter = new MultiTypeAdapter();
    adapter.register(TestItem.class, itemViewBinder);
    assertEquals(1, adapter.getTypePool().size());
    assertEquals(itemViewBinder, adapter.getTypePool().getItemViewBinder(0));

    TestItemViewBinder newBinder = new TestItemViewBinder();
    adapter.register(TestItem.class, newBinder);
    assertEquals(newBinder, adapter.getTypePool().getItemViewBinder(0));
  }


  @Test
  public void shouldNotOverrideRegisteredBinderWhenToMany() {
    MultiTypeAdapter adapter = new MultiTypeAdapter();
    TestItemViewBinder binder2 = new TestItemViewBinder();
    adapter.register(TestItem.class)
        .to(itemViewBinder, binder2)
        .withLinker(new Linker<TestItem>() {
          @Override
          public int index(int position, @NonNull TestItem testItem) {
            // ignored
            return -1;
          }
        });
    assertEquals(TestItem.class, adapter.getTypePool().getClass(0));
    assertEquals(TestItem.class, adapter.getTypePool().getClass(1));
    assertEquals(itemViewBinder, adapter.getTypePool().getItemViewBinder(0));
    assertEquals(binder2, adapter.getTypePool().getItemViewBinder(1));
  }


  @Test
  public void testOnCreateViewHolder() {
    MultiTypeAdapter adapter = new MultiTypeAdapter();
    adapter.register(TestItem.class, mockedItemViewBinder);
    TestItem item = new TestItem("testOnCreateViewHolder");
    adapter.setItems(Collections.singletonList(item));
    int type = adapter.getItemViewType(0);

    adapter.onCreateViewHolder(parent, type);

    verify(mockedItemViewBinder).onCreateViewHolder(inflater, parent);
  }


  @Test
  @SuppressWarnings("deprecation")
  public void testOnBindViewHolder() {
    MultiTypeAdapter adapter = new MultiTypeAdapter();
    adapter.register(TestItem.class, mockedItemViewBinder);
    TestItem item = new TestItem("testOnCreateViewHolder");
    adapter.setItems(Collections.singletonList(item));

    TestItemViewBinder.ViewHolder holder = mock(TestItemViewBinder.ViewHolder.class);
    when(holder.getItemViewType()).thenReturn(adapter.getItemViewType(0));
    adapter.onBindViewHolder(holder, 0);
    verify(mockedItemViewBinder).onBindViewHolder(eq(holder), eq(item), anyList());

    List<Object> payloads = Collections.emptyList();
    adapter.onBindViewHolder(holder, 0, payloads);
    verify(mockedItemViewBinder, times(2)).onBindViewHolder(holder, item, payloads);
  }
}

File: library/src/test/kotlin/me/drakeet/multitype/MultiTypeAdapterTest.kt
/*
 * Copyright (c) 2016-present. Drakeet Xu
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package me.drakeet.multitype

import android.content.Context
import android.view.LayoutInflater
import android.view.ViewGroup
import com.nhaarman.mockitokotlin2.whenever
import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.ArgumentMatchers.anyList
import org.mockito.ArgumentMatchers.anyString
import org.mockito.ArgumentMatchers.eq
import org.mockito.Mock
import org.mockito.Mockito.mock
import org.mockito.Mockito.times
import org.mockito.Mockito.verify
import org.mockito.MockitoAnnotations.initMocks
import org.robolectric.RobolectricTestRunner
import java.util.*

/**
 * @author drakeet
 */
@RunWith(RobolectricTestRunner::class)
class MultiTypeAdapterTest {

  @Mock
  private val parent: ViewGroup? = null
  @Mock
  private val context: Context? = null
  @Mock
  private val mockedItemViewBinder: TestItemViewBinder? = null
  @Mock
  private val inflater: LayoutInflater? = null

  private val itemViewBinder = TestItemViewBinder()

  @Before
  @Throws(Exception::class)
  fun setUp() {
    initMocks(this)
    whenever(parent!!.context).thenReturn(context)
    whenever(context!!.getSystemService(anyString())).thenReturn(inflater)
  }

  @Test
  fun shouldReturnOriginalItems() {
    val list = ArrayList<Any>()
    val adapter = MultiTypeAdapter(list)
    assertEquals(list, adapter.items)
  }

  @Test
  fun shouldReturnEmptyItemsWithDefaultConstructor() {
    val adapter = MultiTypeAdapter()
    assertTrue(adapter.items.isEmpty())
  }

  @Test
  fun shouldOverrideRegisteredBinder() {
    val adapter = MultiTypeAdapter()
    adapter.register(TestItem::class.java, itemViewBinder)
    assertEquals(1, adapter.typePool.size().toLong())
    assertEquals(itemViewBinder, adapter.typePool.getType<Any>(0).binder)

    val newBinder = TestItemViewBinder()
    adapter.register(TestItem::class.java, newBinder)
    assertEquals(newBinder, adapter.typePool.getType<Any>(0).binder)
  }

  @Test
  fun shouldNotOverrideRegisteredBinderWhenToMany() {
    val adapter = MultiTypeAdapter()
    val binder2 = TestItemViewBinder()
    adapter.register(TestItem::class.java)
      .to(itemViewBinder, binder2)
      .withLinker { _, _ -> -1 }
    assertEquals(TestItem::class.java, adapter.typePool.getType<Any>(0).clazz)
    assertEquals(TestItem::class.java, adapter.typePool.getType<Any>(1).clazz)
    assertEquals(itemViewBinder, adapter.typePool.getType<Any>(0).binder)
    assertEquals(binder2, adapter.typePool.getType<Any>(1).binder)
  }

  @Test
  fun testOnCreateViewHolder() {
    val adapter = MultiTypeAdapter()
    adapter.register(TestItem::class.java, mockedItemViewBinder!!)
    val item = TestItem("testOnCreateViewHolder")
    adapter.items = listOf(item)
    val type = adapter.getItemViewType(0)

    adapter.onCreateViewHolder(parent!!, type)

    verify(mockedItemViewBinder).onCreateViewHolder(inflater!!, parent)
  }

  @Test
  fun testOnBindViewHolder() {
    val adapter = MultiTypeAdapter()
    adapter.register(TestItem::class.java, mockedItemViewBinder!!)
    val item = TestItem("testOnCreateViewHolder")
    adapter.items = listOf(item)

    val holder = mock(TestItemViewBinder.ViewHolder::class.java)
    whenever(holder.itemViewType).thenReturn(adapter.getItemViewType(0))
    adapter.onBindViewHolder(holder, 0)
    verify(mockedItemViewBinder).onBindViewHolder(eq(holder), eq(item), anyList())

    val payloads = emptyList<Any>()
    adapter.onBindViewHolder(holder, 0, payloads)
    verify(mockedItemViewBinder, times(2)).onBindViewHolder(holder, item, payloads)
  }
}
